### PR TITLE
fix(vrl): Fix many typedef bugs, introduce `Kind::get` and `Kind::insert`

### PR DESCRIPTION
**Pull request by fuchsnj**
I understand this is a fairly large PR to review, but it is difficult to make larger architecture changes to type definitions without breaking something, so I needed to pull in a decent number of other changes to get to a reasonable stopping point.

closes https://github.com/vectordotdev/vector/issues/13505
closes https://github.com/vectordotdev/vector/issues/13573
closes https://github.com/vectordotdev/vector/issues/13538
closes https://github.com/vectordotdev/vector/issues/13584
closes https://github.com/vectordotdev/vector/issues/13568
closes https://github.com/vectordotdev/vector/issues/13615
closes https://github.com/vectordotdev/vector/issues/13564
closes https://github.com/vectordotdev/vector/issues/13459
closes https://github.com/vectordotdev/vector/issues/13641
closes https://github.com/vectordotdev/vector/issues/11249
closes https://github.com/vectordotdev/vector/issues/13685

Partially fixes
- https://github.com/vectordotdev/vector/issues/13460

### Main Changes
- A new `Kind::insert` function was added. This is the complimentary function for `Value::insert`. It inserts types in the exact same way that the actual value is inserted. The function is infallible. It supports all paths, including coalescing and negative indexing.
It also uses the new lookup code (which also works with the old lookup code). This replaced all uses of `Kind::insert_at_path` which did not support all types of paths, and contained a few bugs.

- A new `Kind::get` function was added, similar to `Kind::insert` which mirrors the functionality of `Value::get`. It is also infallible / uses the new lookup code.

- Both the `Kind::insert` and `Kind::get` functions should now always return the most precise type definition possible. Previously it was common in typedef functions to return `any` as an escape hatch for type definitions that were either unsupported or too complex. If there is a situation were one of these functions returns a type definition that includes a type that can't actually exist, but is possible to model with our current type system, I will consider that a bug.

- Added an `undefined` type. This helps more accurately model certain operations, such as coalescing. VRL does not currently have a concrete `undefined` _value_ type (https://github.com/vectordotdev/vector/issues/13594), but `undefined` can still appear transiently in some situations, such as an out of bounds index access on an array, or accessing a field on an object that doesn't exist. Accessing an undefined value implicitly converts the value to `null`, which is correctly modeled here. This PR doesn't attempt to use `undefined` in _all_ possible cases yet, it just introduces the majority of changes needed.

### CODE REVIEW

File: lib/vrl/tests/tests/diagnostics/unnest_typedefs.vrl
Comment (fuchsnj) [inline lib/vrl/tests/tests/diagnostics/unnest_typedefs.vrl:2]: The removed assertion here were asserting invalid coalescing behavior. There are now unit tests asserting the correct behavior.

File: lib/value/src/kind/merge.rs
Comment (spencergilbert) [inline lib/value/src/kind/merge.rs:251]: This extra line break, why?!

File: lib/value/src/kind/merge.rs
Comment (spencergilbert) [inline lib/value/src/kind/merge.rs:42]: Is this a breaking change?

File: website/content/en/highlights/2022-08-16-0-24-0-upgrade-guide.md
Comment (spencergilbert) [inline website/content/en/highlights/2022-08-16-0-24-0-upgrade-guide.md:113]:
```suggestion
For example, if you are trying to `split` a string, but the input could now be null, the error would look like this
```

**Review by spencergilbert — COMMENTED** [review]
Few observations/questions - I haven't read through the new `crud` files and I wanted to take a second pass through the PR as well.

File: lib/value/src/kind/crud/mod.rs
Comment (fuchsnj) [reply lib/value/src/kind/crud/mod.rs:2]: we have the **R**ead (get) and **U**pdate (insert). I plan on adding delete (remove) soon, in a different PR.

File: lib/value/src/kind/merge.rs
Comment (fuchsnj) [reply lib/value/src/kind/merge.rs:42]: Nope. The entire `merge` function is deprecated and I've been removing pieces of it. This has been one of the main root causes of some bugs. This PR is removing the "Append" strategy and has moved that functionality to other functions specifically for that.

File: lib/value/src/kind.rs
Comment (spencergilbert) [reply lib/value/src/kind.rs:111]: Np, just making sure my reading was right - not super used to seeing the double `self` usage 🙂

File: lib/value/src/kind/builder.rs
Comment (spencergilbert) [inline lib/value/src/kind/builder.rs:43]: Was this not used before? I didn't notice an equivalent replacement.

File: lib/vrl/stdlib/src/unnest.rs
Comment (spencergilbert) [inline lib/vrl/stdlib/src/unnest.rs:196]: Is this a TODO that we can eventually resolve?

File: lib/value/src/kind/crud/get.rs
Comment (spencergilbert) [inline lib/value/src/kind/crud/get.rs:11]:
```suggestion
    /// the implicit conversion of "undefined" to "null".
```

File: lib/value/src/kind/crud/get.rs
Comment (spencergilbert) [inline lib/value/src/kind/crud/get.rs:104]: Intended to be included? Two more `debug_assert!()`s in the insert file as well.

File: lib/value/src/kind/crud/get.rs
Comment (spencergilbert) [inline lib/value/src/kind/crud/get.rs:168]: With the number of test cases, it feels like it'd be excellent to be able to do this programmatically with proptest.

File: lib/value/src/kind/crud/get.rs
Comment (fuchsnj) [reply lib/value/src/kind/crud/get.rs:104]: Yes, those are to verify correctness. These are disabled in release builds.

File: lib/vrl/stdlib/src/unnest.rs
Comment (fuchsnj) [reply lib/vrl/stdlib/src/unnest.rs:196]: Yes, once the linked issue is fixed we should be able to change this.

File: lib/value/src/kind/builder.rs
Comment (fuchsnj) [reply lib/value/src/kind/builder.rs:43]: This was not used. Originally many similar methods were added to `Kind` that weren't used yet, assuming they would be useful. We are pretty far into using `Kind` now and we haven't seen a use for this yet, so I figured it was time to remove it.
I was originally trying to decide if `undefined` would be considered a "primitive" or not, and didn't see an example usage.

File: lib/value/src/kind/crud/get.rs
Comment (spencergilbert) [reply lib/value/src/kind/crud/get.rs:104]: I read the doc page for it, but totally skipped over that section somehow. 👍

File: lib/value/src/kind/crud/get.rs
Comment (fuchsnj) [reply lib/value/src/kind/crud/get.rs:168]: I plan on adding prop testing for typedefs in the future. I think these are useful in addition to that though.

File: lib/value/src/kind/comparison.rs
Comment (JeanMertz) [inline lib/value/src/kind/comparison.rs:16]: Are we sure we want `any` to be defined as "any type, including not being defined at all"?

It almost feels like modelling this properly would require two levels, first you'd have an `undefined` vs `defined` state for a `Kind`, and then for defined ones, you'd have all the different type states that kind can have.

That is, there are two separate questions you'd want to be able to ask about the state of a value:

1. Do we know if the value can exist at runtime?
2. If it can exist, what do we know about the potential types of the value?

In this implementation, we're merging those two questions, which (it seems to me, but I might be off here) muddies the water when it comes to methods like `any`, which no longer implies "any type", but "any type _and_ the potential for the value to not exist at all".

I haven't thought this through thoroughly yet, but this _feels_ like it would complicate the reasoning we need to do on type definitions.

I'm curious about your thoughts here, and also happy to chat offline.

File: lib/value/src/kind/comparison.rs
Comment (JeanMertz) [inline lib/value/src/kind/comparison.rs:219]: This is another one in which its outcome is muddied by the current API.

If I have a value, that can either exist and be a string, or not exist at all, I consider the type of that value to be "exact" _when the value exists_. When the value doesn't exist, there's no concept of "exactness" of the type, since there's no type to tie to a non-existing value.

File: lib/value/src/kind/merge.rs
Comment (JeanMertz) [inline lib/value/src/kind/merge.rs:42]: Probably good to file an issue to remove the use of this method in a follow-up PR.

**Review by JeanMertz — CHANGES_REQUESTED** [review]
Overall this looks good to me, but I do have some worries/questions about muddying the `Kind` API with the `undefined` state, specifically the fact that it is modelled similarly to the actual type states, even though it seems to me that it serves a different purpose.

I'd love to talk this through first. It also looks like most bugs that were fixed happened by switching to the new `get`/`insert` logic, so I do question the need to introduce the `undefined` state without an RFC first (or at least a broader discussion on the direction) in this PR, which now blocks a bunch of other fixes, only increasing the want to get this merged sooner rather than later, leaving little room for higher-level discussions that don't impact the bug fixes.

File: lib/value/src/kind/collection.rs
Comment (fuchsnj) [reply lib/value/src/kind/collection.rs:109]: I'll add a comment, but today either `never` or `undefined` would be correct. I'd like the "canonicalized" form to eventually be just "undefined" since that's the _actual_ type of types that aren't defined with a closed type.

File: lib/value/src/kind/collection.rs
Comment (fuchsnj) [reply lib/value/src/kind/collection.rs:151]: Yes, I think this one got lost. I'll add an issue with more info, but as far as I can tell this has always been broken (at least since 0.19.0) and was refactored a few times keeping the bug. I'll link the bug soon with an example if you are more curious, and add a test.

Comment (fuchsnj) [reply lib/value/src/kind/collection.rs:151]: https://github.com/vectordotdev/vector/issues/13685

File: lib/value/src/kind/comparison.rs
Comment (fuchsnj) [reply lib/value/src/kind/comparison.rs:16]: > Are we sure we want any to be defined as "any type, including not being defined at all"?

Yes. `any` means we have no type information (it can be anything). If you know it can be anything except undefined, you can set it to `Kind::any().without_undefined()`.
I think you are assuming `undefined` is more special than it is. It's just another type a value can be, and happens to be the default type for values not defined.

File: lib/value/src/kind/comparison.rs
Comment (fuchsnj) [reply lib/value/src/kind/comparison.rs:219]: > When the value doesn't exist, there's no concept of "exactness" of the type
It is very cleanly defined as "at most one type is set". `undefined` is a normal type, similar to `null`.

**Comment by fuchsnj** [comment]
@JeanMertz It seems like your main concern is around the introduction of the `undefined` type.

>I do have some worries/questions about muddying the Kind API with the undefined state, specifically the fact that it is modelled similarly to the actual type states, even though it seems to me that it serves a different purpose.

I consider the `undefined` state to be an "actual" type state. It serves the same purpose as all the other types states. It simply models the fact that a value could potentially be undefined. This may be more confusing since VRL today doesn't currently have a concrete `undefined` _value_, but undefined values still pop up transiently in some places, such as out of bounds array indices, and accessing fields of an object that aren't defined.

There is a tracking issue for discussion around adding the undefined value here: https://github.com/vectordotdev/vector/issues/13594

**Comment by JeanMertz** [comment]
> I consider the `undefined` state to be an "actual" type state. It serves the same purpose as all the other types states. It simply models the fact that a value could potentially be undefined. This may be more confusing since VRL today doesn't currently have a concrete `undefined` _value_, but undefined values still pop up transiently in some places, such as out of bounds array indices, and accessing fields of an object that aren't defined.

Right. I get that this is how this PR currently works, but I'm saying I'm not entirely sold on that solution going forward.

I'll also mention straight-away that I don't have an alternative prepared for this discussion, but I'd rather see us write up several alternative solutions, their pro's and cons, and then decide on the direction to take. It could be that you already did this for yourself, if so, then maybe just writing down those alternatives and their characteristics would be enough to resolve this discussion, assuming this variant is indeed the one we favour the most.

> There is a tracking issue for discussion around adding the undefined value here: #13594

I am aware, but at least to me, a "tracking issue" is to track work we agreed needs to happen, which either exists because it's a bug we know how to fix, or a feature we've settled a direction on (most likely through an RFC, or some other form of informal discussions for smaller features). The fact that this issue has the "needs approval" tag, signals to me that we're not yet ready to implement an architectural change such as this, based on the existence of that issue alone.

Again, I don't want to be pedantic, and I don't particularly mind experimentation, I just am not (yet) convinced this is the right approach we should take, and having all the work done before we are aligned on the approach could potentially be wasteful and demotivating for whoever did the work, if we do decide to take a different approach.

The process of discussing and/or writing RFCs (or have informal discussions, and writing the outcome of those discussions in an issue) before we start implementing is not to introduce unnecessary red-tape, but to avoid situations where unplanned work turns out to go in a different direction than we'd like to go, and in this case, potentially holding up other planned work (or at least other work we _do_ want to get in).

Let's chat offline. As I said, it might just be a factor of writing up the alternatives and concluding this is still the direction we should go.

**Comment by fuchsnj** [comment]
> I'll also mention straight-away that I don't have an alternative prepared for this discussion, but I'd rather see us write up several alternative solutions, their pro's and cons, and then decide on the direction to take. It could be that you already did this for yourself, if so, then maybe just writing down those alternatives and their characteristics would be enough to resolve this discussion, assuming this variant is indeed the one we favour the most.

I'm not aware of other reasonable solutions to tracking the `undefined` state, as I thought this was a fairly straightforward extension of our current type definition model.  I'd be happy to discuss alternatives if you can think of any though.

> I am aware, but at least to me, a "tracking issue" is to track work we agreed needs to happen, which either exists because it's a bug we know how to fix, or a feature we've settled a direction on (most likely through an RFC, or some other form of informal discussions for smaller features). The fact that this issue has the "needs approval" tag, signals to me that we're not yet ready to implement an architectural change such as this, based on the existence of that issue alone.

I think I worded this wrong, you are right It's not a "tracking issue". I want to clarify that this issue is for discussing potential future work of adding a concrete `undefined` _value_, and is separate and not required for this PR which added an `undefined` _type_ to model already existing behavior.

**Review by JeanMertz — APPROVED** [review]
As discussed offline, while the path that got us here could be better, this is still an improvement over the current situation, and given that, we'll move forward as-is, with the possibility of refactoring this as-needed in the future.

File: website/content/en/highlights/2022-08-16-0-24-0-upgrade-guide.md
Comment (jszwedko) [inline website/content/en/highlights/2022-08-16-0-24-0-upgrade-guide.md:100]: 👍 thanks for adding this note.

**Pull request merged**
